src/gridcoin/tally.cpp

```diff
@@ -322,17 +390,52 @@ class ResearcherTally
             account.m_total_magnitude -= pindex->Magnitude();
         }
 
-        pindex = pindex->pprev;
+        pindex = FindLastRewardBlock(cpid, pindex);
 
-        while (pindex
-            && (pindex->ResearchSubsidy() <= 0 || pindex->GetMiningId() != cpid))
-        {
-            pindex = pindex->pprev;
-        }
+        account.m_last_block_ptr = pindex;
+    }
 
-        assert(pindex && pindex != pindexGenesisBlock);
+    //!
+    //! \brief Disassociate a block's MRC research reward data from the tally.
+    //! \param pindex Contains information about the block to erase.
+    //!
+    void ForgetMRCRewardBlock(const CBlockIndex* pindex)
+    {
+        for (const auto& mrc_researcher : pindex->m_mrc_researchers) {
+            Cpid cpid = mrc_researcher->m_cpid;
 
-        account.m_last_block_ptr = pindex;
+            auto iter = m_researchers.find(cpid);
+
+            assert(iter != m_researchers.end());
+
+            ResearchAccount& account = iter->second;
+
+            assert(account.m_first_block_ptr != nullptr);
+            assert(pindex->pprev == account.m_last_block_ptr);
+
+            // MRC's are paid on the last block prior to the staked block (i.e. pprev), but recorded in the pindex
+            // where the claim is. This is a little tricky.
+            //
+            // When disconnecting a CPID's first block, reset the account, but
+            // retain the pending snapshot accrual amount:
+            //
+            if (pindex->pprev == account.m_first_block_ptr) {
+                account = ResearchAccount(account.m_accrual);
+                return;
+            }
+
+            account.m_total_research_subsidy -= mrc_researcher->m_research_subsidy;
+
+            if (mrc_researcher->m_magnitude > 0) {
+                account.m_accuracy--;
+                account.m_total_magnitude -= pindex->pprev->Magnitude();
+            }
+
+            // This search is based on the contents of pindex and so takes and outputs pindex as the argument even for MRCs.
+            pindex = FindLastRewardBlock(cpid, pindex);
+
+            account.m_last_block_ptr = pindex->pprev;
```

Comment:
```suggestion
            account.m_last_block_ptr = pindex;
```
Missed this, the other two comments were just to show to clarify the point.

Reply:
Wait a second. Isn't that just a shell game? whether you put the step to pprev here or in the search, doesn't it result in the same thing?

Reply:
The issue you were describing earlier is precisely why I had the m_last_block_ptr set to pindex->pprev here. One of us is confused, and I am not sure whom at this point.

Reply:
Well... I think we played a shell game here, but applying your suggested commit above fixes it.

Reply:
> whether you put the step to pprev here or in the search, doesn't it result in the same thing?

Not exactly.

Let's assume a chain consisting of Stake(0) -> NOP(1) -> MRC(2) -> NOP(3) -> Stake(4) -> NOP(5) -> MRC(6) for a researcher. When the chain rollbacks one block(6), ForgetMRCRewardBlock is triggered and calls FindLastRewardBlock. This method would return the block number 4 in this case. If the ForgetMRCRewardBlock used pprev here, it would mistakenly refer to 3 instead of 4 because of the pprev.

Going a step further and rolling back two more blocks(fifth block is not relevant), rolling back number 4 would trigger ForgetRewardBlock. The function itself would call FindLastRewardBlock which would return 2. Since MRC payments are based off on the last block, the actual height should've been 1 but it's missing a pprev.

Moving the logic to FindLastRewardBlock ensures that, in the first case the same number is returned. ForgetMRCRewardBlock does not have to do a pprev, so the height is correctly 4. In the second case, the function would return 1 correctly and no modification would be needed to do for ForgetRewardBlock.

Reply:
I'll add a unit test for this, so you can check.